### ISSUE TITLE
Starter HTTP Message headers overwritten when using Message Value Setter: Request Body

### ISSUE THREAD

**Issue by tomdottom** [synfron/ReshaperForBurp#41]
# Steps to reproduce

1. Create a _Starter HTTP Message_ as a global variable with the contents:

```
POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close
```

![image](https://github.com/synfron/ReshaperForBurp/assets/5106166/b12ecf44-f266-4275-9830-4404433b3b63)

2. Create a new _Rule_  with a _Then_ using `Build HTTP Message: Request`
    - Use the the starter message from step 1.
    - Configure a `Message Value Setter` which sets the `Request Body` to some simple text like _Test_
    - Set the destination variable

3. Add a _Then_ to log the request variable from step 2.

![image](https://github.com/synfron/ReshaperForBurp/assets/5106166/b87155d7-e913-4315-8cf7-86a5bae4b3f3)

4. Send a request through Burp to trigger the _Rule_ then look in the logs.  It should be apparent that the `Host` and `Content-Type` headers from the starter message have been removed:

```
POST /post HTTP/1.1
Content-Length: 4

Test
```

![image](https://github.com/synfron/ReshaperForBurp/assets/5106166/9587389d-eee5-43b4-9ead-3631b9f8eb83)

# Rule export

```yaml
rules:
- whens:
  - !<.WhenEventDirection>
    negate: false
    useOrCondition: false
    dataDirection: "Request"
  thens:
  - !<.ThenBuildHttpMessage>
    dataDirection: "Request"
    starterHttpMessage: "{{g:TestHttpPostRequest}}"
    messageValueSetters:
    - destinationMessageValue: "HttpRequestBody"
      destinationIdentifier: ""
      destinationIdentifierPlacement: "Only"
      sourceText: "Test"
    destinationVariableSource: "Event"
    destinationVariableName: "TestRequestBody"
  - !<.ThenLog>
    text: "{{e:TestRequestBody}}"
  enabled: true
  autoRun: true
  name: "Test Dissappearing Post Headers"
webSocketRules: []
variables: []
```

**Comment by ddwightx** [comment]
Very interesting. It is not reproducing for me. What version of Reshaper do you have?

![image](https://github.com/synfron/ReshaperForBurp/assets/48854453/188fe900-05eb-46d1-9a07-43db55c876b7)
![image](https://github.com/synfron/ReshaperForBurp/assets/48854453/b3f91102-abce-4add-9ba3-e5dc6f96f692)
![image](https://github.com/synfron/ReshaperForBurp/assets/48854453/b31a4c24-2c69-4077-9815-bf10323c7e87)
![image](https://github.com/synfron/ReshaperForBurp/assets/48854453/4fc71046-4bc3-4da3-b3a7-e6a108ec6b11)

**Comment by tomdottom** [comment]
> "extension_file":"bapps/7bcec7656b5746e9a85c427f243e6d5a/build/libs/reshaper-for-burp-2.1.0.jar"

**Comment by ddwightx** [comment]
On the Settings tab, check the box for `Enable Event Diagnostics` trigger the rule again, then go to the Log tab and copy the output here.

Mine looks like this:
```
{
  "message" : "POST /post HTTP/1.1\r\nHost: 0.0.0.0:9999\r\nContent-Type: application/json\r\nConnection: close\r\nContent-Length: 4\r\n\r\nTest"
}
Request: https://example.com/
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Request' equals 'Request') - PASS
				    Then Build HTTP Message(dataDirection='Request' starterHttpMessage='POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close' destinationVariableSource='Event' destinationVariableName='TestRequestBody' {{message:httprequestbody}}='Test') 
				    Then Log('POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close
Content-Length: 4

Test') 
	End Rule
End Request

Response: https://example.com/
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Response' equals 'Request') - FAIL
	End Rule
End Response
```

**Comment by tomdottom** [comment]
```
{
  "message" : "POST /post HTTP/1.1\r\nContent-Length: 4\r\n\r\nTest"
}
Request: https://api.alpha-e2e.icaas.theonenetwork.ericsson.net/icaas-user-mgmt/v1/users/64d2dae0d2c89b4767fdc444/enable-user-at-login
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Request' equals 'Request') - PASS
				    Then Build HTTP Message(dataDirection='Request' starterHttpMessage='POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close

' destinationVariableSource='Event' destinationVariableName='TestRequestBody' {{message:httprequestbody}}='Test') 
				    Then Log('POST /post HTTP/1.1
Content-Length: 4

Test') 
	End Rule
End Request

Response: https://api.alpha-e2e.icaas.theonenetwork.ericsson.net/icaas-user-mgmt/v1/users/64d2dae0d2c89b4767fdc444/enable-user-at-login
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Response' equals 'Request') - FAIL
	End Rule
End Response
```

**Comment by ddwightx** [comment]
That's very odd. Can you try a reinstall?
Otherwise, if you wish, you can try the release here by using the jar provided: https://github.com/synfron/ReshaperForBurp/releases/tag/v2.1.0
You'll need to disable the BAppStore version before installing that one. Otherwise they will conflict.

**Comment by tomdottom** [comment]
Uninstalled & reinstalled from BApp Store - Same behaviour.
Uninstalled & installed Jar from above link - Same behaviour.

**Comment by tomdottom** [comment]
Note that I changed the setter to `Request URI Path` and still see the headers getting removed
```
{
  "message" : "POST /Test HTTP/1.1\r\n\r\n"
}
Request: https://www.google.com/
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Request' equals 'Request') - PASS
				    Then Build HTTP Message(dataDirection='Request' starterHttpMessage='POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close

' destinationVariableSource='Event' destinationVariableName='TestRequestBody' {{message:httprequesturipath}}='Test') 
				    Then Log('POST /Test HTTP/1.1

') 
	End Rule
End Request

Response: https://www.google.com/
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Response' equals 'Request') - FAIL
	End Rule
End Response
```

**Comment by tomdottom** [comment]
I just tried to run the _Then_ `Parse HTTP Message: Request` on the starter HTTP message_ and can see a parsing error in the logs.

![image](https://github.com/synfron/ReshaperForBurp/assets/5106166/0c9ac720-81ed-4466-91b8-94716dd92055)

```
{
  "message" : "Starter Message: POST /post HTTP/1.1\nHost: 0.0.0.0:9999\nContent-Type: application/json\nConnection: close\n\n"
}
{
  "message" : "POST /Test HTTP/1.1\r\n\r\n"
}
{
  "message" : "null"
}
Request: https://www.google.com/
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Request' equals 'Request') - PASS
				    Then Log('Starter Message: POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close

') 
				    Then Build HTTP Message(dataDirection='Request' starterHttpMessage='POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close

' destinationVariableSource='Event' destinationVariableName='TestRequestBody' {{message:httprequesturipath}}='Test') 
				    Then Log('POST /Test HTTP/1.1

') 
				    Then Parse HTTP Message('[]') - ERRORED
				    Then Log('null') 
	End Rule
End Request

Response: https://www.google.com/
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Response' equals 'Request') - FAIL
	End Rule
End Response
```

**Comment by tomdottom** [comment]
Just the defaults which came when installing

![image](https://github.com/synfron/ReshaperForBurp/assets/5106166/a5739bb0-0ca0-4277-a7d3-5dfc5171edf1)

**Comment by tomdottom** [comment]
> I have that fixed in https://github.com/synfron/ReshaperForBurp/releases/tag/v2.1.0-next-v2 which I haven't officially released yet.

Installed `v2.1.0-next-v2 ` and still don't see the headers.  However, I no longer see the parse error.

```
{
  "message" : "POST /post HTTP/1.1\r\nContent-Length: 4\r\n\r\nTest"
}
Request: https://www.google.com/
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Request' equals 'Request') - PASS
				    Then Build HTTP Message(dataDirection='Request' starterHttpMessage='POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close

' destinationVariableSource='Event' destinationVariableName='TestRequestBody' {{message:httprequestbody}}='Test') 
				    Then Log('POST /post HTTP/1.1
Content-Length: 4

Test') 
				    Then Parse HTTP Message(dataDirection='Request' httpMessage='POST /post HTTP/1.1
Host: 0.0.0.0:9999
Content-Type: application/json
Connection: close

' {{event:tmp}}='') 
	End Rule
End Request

Response: https://www.google.com/
	Rule: Test Dissappearing Post Headers
				    When Event Direction('Response' equals 'Request') - FAIL
	End Rule
End Response
```

**Comment by tomdottom** [comment]
Gtg.  

I'll test this on some other Burp installations over the coming days and make note where I'm seeing  working and broken behaviour.

- T

**Comment by ddwightx** [comment]
@tomdottom Thanks for your efforts so far! Unfortunately, I'm lost on this one so far.

**Comment by ddwightx** [comment]
@tomdottom Good new. I downloaded Arch Linux and am able to reproduce the issue 😊.

**Comment by tomdottom** [comment]
That's great news.  Is it an Arch issue, or a general Linux issue?

**Comment by ddwightx** [comment]
It's actually a CRLF versus LF only issue. The spec for headers is that each header line must end in a CRLF. On Linux, the CR's are not there when copy and pasting your example starter request.  On Windows, they are there. Will need to consider the scenario with no CR.

**Comment by tomdottom** [comment]
🤦 of course.  Should have thought of that.

**Comment by ddwightx** [comment]
Reshaper is using Burp's library to get the headers from the request (`HttpRequest.httpRequest(ByteArray.byteArray(request))`). That returns no headers when LF is used alone. Fixing this would require not using Burp's library which will carry its own risks.

**Comment by tomdottom** [comment]
I would suggest just changing the editor. 

If I could have seen `\r\n` when editing the stater http template I may have realised what was going on. 

Or maybe there's a way to use Burps message editor?

**Comment by ddwightx** [comment]
Good suggestion. I'll see what Burp's message editor offers.

**Comment by ddwightx** [comment]
Switched to use the Burp message editor in https://github.com/synfron/ReshaperForBurp/releases/tag/v2.1.0-next-v3

**Comment by ddwightx** [comment]
The issue is addressed in [Reshaper for Burp v2.2.0](https://github.com/synfron/ReshaperForBurp/releases/tag/v2.2.0). The Global Variables UI has been switched to use Burp's message editor which has options to view CR and LF characters. Also added a sanity check log to warn if the missing CR situation is detected. 

 It will be some time for the release to reach the BApp Store because of their review process.
